packages/@aws-cdk/aws-ec2/lib/user-data.ts

```diff
@@ -0,0 +1,143 @@
+import cdk = require('@aws-cdk/cdk');
+import { CloudFormationInit } from './cloudformation-init';
+import { OperatingSystem, UserDataOptions } from './machine-image';
+
+/**
+ * Properties for creating a UserData
+ */
+export interface UserDataProps extends UserDataOptions {
+  /**
+   * Operating system to generate UserData for
+   */
+  os: OperatingSystem;
+
+  /**
+   * Default resource to generate signals for
+   */
+  defaultResource?: cdk.Resource;
+}
+
+/**
+ * Class that represents user data
+ */
+export class UserData extends cdk.Construct {
+  private readonly lines = new Array<string>();
+  private readonly os: OperatingSystem;
+
+  constructor(parent: cdk.Construct, id: string, private readonly props: UserDataProps ) {
+    super(parent, id);
+    this.os = props.os;
+  }
+
+  /**
+   * Add a literal command to the userdata
+   */
+  public addCommand(...lines: string[]) {
+    this.lines.push(...lines);
+  }
+
+  /**
+   * Add a command to signal a resource
+   */
+  public addResourceSignalCommand(options: ResourceSignalOptions = {}) {
+    // https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/cfn-signal.html
+    const resource = options.resource || this.props.defaultResource;
+    if (!resource) {
+      throw new Error('UserData does not have a default resource; please provide an explicit resource to signal');
+    }
+
+    const parts = ['cfn-signal',
+      `--region ${new cdk.AwsRegion()}`,
+      `--stack ${new cdk.AwsStackId()}`,
+      `--resource ${resource.logicalId}`
```

Comment:
This should be the EC2 instance that you want to signal

Reply:
Or AutoScalingGroup.

Reply:
We have this in quite a few userdata sections

```
            # Ensure our PATH is set correctly (on Amazon Linux, cfn-signal is in /opt/aws/bin)
            . ~/.bash_profile
```
This has served us well, but perhaps we want to offer a path override for the location of cfn-signal?

Reply:
Feels like we should hardcode the path in the script (w/ cdk override)